python/ray/air/dataset_iterator.py

```diff
@@ -0,0 +1,168 @@
+import abc
+from typing import TYPE_CHECKING, Dict, List, Optional, Union, Iterator
+from typing_extensions import Literal
+
+from ray.air.data_batch_type import DataBatchType
+from ray.util.annotations import PublicAPI
+
+if TYPE_CHECKING:
+    import tf
+    import torch
+
+
+TorchTensorBatchType = Union["torch.Tensor", Dict[str, "torch.Tensor"]]
+
+
+@PublicAPI(stability="beta")
+class DatasetIterator(abc.ABC):
+    """An iterator for reading a preprocessed dataset.
+
+    Each iteration call represents one pass (epoch) over the dataset. During
+    training, each trainer actor should get its own iterator by calling
+    `session.get_dataset_shard("train")`.
+
+    Tip: For debugging purposes, use
+    :py:meth:`~ray.air.util.check_ingest.make_local_dataset_iterator`
```

Comment:
Nit: you don't need `:py:` because Sphinx assumes it beydefault

```suggestion
    :meth:`~ray.air.util.check_ingest.make_local_dataset_iterator`
```